kbmadm/kbmadm.c

```diff
@@ -0,0 +1,1123 @@
+	    (ret = add_guidstr(req, KBM_NV_GUID, guidstr)) != ERRF_OK)
+		goto done;
+
+	if (template_f != NULL &&
+	    (ret = add_template_file(req, KBM_NV_TEMPLATE,
+	    template_f)) != ERRF_OK)
+		goto done;	
+
+	if ((ret = envlist_add_string(req, KBM_NV_DATASET,
+	    dataset)) != ERRF_OK)
+		goto done;
+
+	if ((ret = send_request(req, &resp)) != ERRF_OK)
+		goto done;
+
+	/*
+	 * Build zpool create command line:
+	 * 'zpool create <options from kbmd>'
+	 */
+	if ((ret = strarray_append(&args, "zpool")) != ERRF_OK ||
+	    (ret = strarray_append(&args, "create")) != ERRF_OK ||
+	    (ret = add_create_args(&args, resp)) != ERRF_OK)
+		goto done;
+
+	/*
+	 * Append arguments from kbmadm create-zpool command line.
+	 * argv[] looks similar to:
+	 *	create-zpool
+	 *	arg1
+	 *	arg2
+	 *	...
+	 * So we start at arg[1].
+	 */
+	for (size_t i = 1; i < argc; i++) {
+		if ((ret = strarray_append(&args, "%s", argv[i])) != ERRF_OK)
+			goto done;
+	}
+
+	if ((ret = envlist_lookup_uint8_array(resp, KBM_NV_ZPOOL_KEY, &key,
+	    &keylen)) != ERRF_OK)
+		goto done;
+
+	ret = run_zpool_cmd(args.sar_strs, key, keylen);
+
+done:
+	nvlist_free(req);
+	nvlist_free(resp);
+	strarray_fini(&args);
+	return (ret);
+}
+
+static errf_t *
+run_zpool_cmd(char **argv, const uint8_t *key, size_t keylen)
+{
+	errf_t *ret = ERRF_OK;
+	custr_t *out[2] = { 0 };
+	int fds[3] = { -1, STDOUT_FILENO, STDERR_FILENO };
+	int status;
+	pid_t pid;
+
+	if ((ret = spawn(ZPOOL_CMD, argv, _environ, &pid, fds)) != ERRF_OK ||
+	    (ret = interact(pid, fds, key, keylen, out, &status,
+	    B_FALSE)) != ERRF_OK)
+		return (ret);
+
+	if (status != 0) {
+		return (errf("CommandError", NULL, "zpool create returned %d",
+		    status));
+	}
+
+	return (ERRF_OK);
+}
+
+static errf_t *
+unlock_dataset(const char *dataset)
+{
+	errf_t *ret = ERRF_OK;
+	nvlist_t *req = NULL;
+	nvlist_t *resp = NULL;
+
+	if ((ret = req_new(KBM_CMD_ZFS_UNLOCK, &req)) != ERRF_OK ||
+	    (ret = envlist_add_string(req, KBM_NV_ZFS_DATASET,
+	    dataset)) != ERRF_OK) {
+		goto done;
+	}
+
+	ret = send_request(req, &resp);
+
+done:
+	nvlist_free(req);
+	nvlist_free(resp);
+	return (ret);
+}
+
+static errf_t *
+do_unlock(int argc, char **argv)
+{
+	errf_t *ret;
+	int c;
+	boolean_t recover_if_needed = B_FALSE;
+
+	while ((c = getopt(argc, argv, "r")) != -1) {
+		switch (c) {
+		case 'r':
+			recover_if_needed = B_TRUE;
+			break;
+		case '?':
+			(void) fprintf(stderr, "Unknown option -%c\n",
+			    optopt);
+			usage();
+		}
+	}
+
+	if (argv[optind] == NULL) {
+		return (errf("ArgumentError", NULL, "dataset name is missing"));
+	}
+
+	for (int i = 1; i < argc; i++) {
+		ret = unlock_dataset(argv[i]);
+		if (ret != ERRF_OK) {
+			if (errf_caused_by(ret, "RecoveryNeeded") &&
+			    recover_if_needed) {
+				warnfx(ret, "");
+				errf_free(ret);
+
+				ret = recover(argv[i], 0);
+			}
+
+			if (ret != ERRF_OK) {
+				goto done;
+			}
+		}
+
+		if (IS_ZPOOL(argv[i])) {
+			mount_zpool(argv[i], NULL);
+		}
```

Comment:
The mounting of all datasets in a pool (legacy and canmount=noauto excepted) when a pool is imported had a sufficiently different code path than `zfs mount -a` that I couldn't convince myself that performing `zfs mount -a` after importing and unlocking all the pools would be 100% equivalent to what happens today. Having been burned enough with just trying to have the admin network come up earlier in the boot process (and all the corner cases that came up), I wanted to keep the process here as close to the existing one as possible to try to avoid any problems.  Today we import any pools present one at a time, which then mounts any non-legacy and any canmount=yes datasets after each pool is imported.  After this, when we import a pool, we unlock the pool, then do the same mount operation that was previously done during `zpool import $pool`, so that we have the same net effect.

Reply:
ok, if needed we can always extend this with a "do not mount" flag.

Reply:
confirmed